### ISSUE TITLE
Show more useful information in the "Associated Stations" table

### ISSUE THREAD

**Issue by mauermann** [openwrt/luci#540]
It would be nice if the "Associated Stations" table would show more useful information. e.g. with dual band radios getting more common it would be interesting to see the radio a station is associated. Also, instead of remembering MAC addresses, I think it's better to see its hostname or IP address.

Attached screenshot shows my proposed changes with columns for hostname and radio added. Unfortunately because of table width I had to drop the Noise column. However, this is less important for me.

I'm sure it needs some work and discussion, so it's probably too early to submit a pull request... ;-)

![bildschirmfoto vom 2015-11-05 12 31 49](https://cloud.githubusercontent.com/assets/5954089/10967822/8ef4e578-83bd-11e5-9904-bd5e7af559b0.png)

**Comment by mauermann** [comment]
Here is my diff for the screenshot above. It's currently only for status/overview, but should be possible for network/wireless, too.

```diff
--- view/admin_status/index.htm.bak	2015-11-05 09:52:41.000000000 +0100
+++ view/admin_status/index.htm	2015-11-05 13:35:07.000000000 +0100
@@ -403,6 +403,18 @@
 
 						for (var bssid in net.assoclist)
 						{
+							var macaddr = bssid.toLowerCase();
+							var host = '?';
+							<% if has_dhcp then %>
+							for (var lidx = 0; lidx < info.leases.length; lidx++)
+							{
+								if (info.leases[lidx].macaddr == macaddr) {
+									host = info.leases[lidx].hostname ? info.leases[lidx].hostname : info.leases[lidx].ipaddr;
+									break;
+								}
+							}
+							<% end %>
+							
 							assoclist.push({
 								bssid:    bssid,
 								signal:   net.assoclist[bssid].signal,
@@ -414,7 +426,9 @@
 								tx_mcs:   net.assoclist[bssid].tx_mcs,
 								tx_40mhz: net.assoclist[bssid].tx_40mhz,
 								link:     net.link,
-								name:     net.name
+								name:     net.name,
+								device:   dev.device,
+								host:     host
 							});
 						}
 					}
@@ -464,23 +478,24 @@
 					);
 
 					tr.insertCell(-1).innerHTML = assoclist[i].bssid;
+					tr.insertCell(-1).innerHTML = assoclist[i].host.length > 18 ? assoclist[i].host.substring(0,15)+"..." : assoclist[i].host;
 
 					tr.insertCell(-1).innerHTML = String.format(
 						'<a href="%s">%s</a>',
 							assoclist[i].link,
 							'%h'.format(assoclist[i].name).nobr()
 					);
+					tr.insertCell(-1).innerHTML = assoclist[i].device;
 
 					tr.insertCell(-1).innerHTML = String.format('%d <%:dBm%>', assoclist[i].signal).nobr();
-					tr.insertCell(-1).innerHTML = String.format('%d <%:dBm%>', assoclist[i].noise).nobr();
 
 					tr.insertCell(-1).innerHTML = (assoclist[i].rx_mcs > -1)
-						? String.format('%.1f <%:Mbit/s%>, MCS %d, %d<%:MHz%>', assoclist[i].rx_rate / 1000, assoclist[i].rx_mcs, assoclist[i].rx_40mhz ? 40 : 20).nobr()
+						? String.format('%.1f <%:Mbit/s%>, MCS %d, %d<%:MHz%>', assoclist[i].rx_rate / 1000, assoclist[i].rx_mcs, assoclist[i].rx_40mhz ? 40 : 20)
 						: String.format('%.1f <%:Mbit/s%>', assoclist[i].rx_rate / 1000).nobr()
 					;
 
 					tr.insertCell(-1).innerHTML = (assoclist[i].tx_mcs > -1)
-						? String.format('%.1f <%:Mbit/s%>, MCS %d, %d<%:MHz%>', assoclist[i].tx_rate / 1000, assoclist[i].tx_mcs, assoclist[i].tx_40mhz ? 40 : 20).nobr()
+						? String.format('%.1f <%:Mbit/s%>, MCS %d, %d<%:MHz%>', assoclist[i].tx_rate / 1000, assoclist[i].tx_mcs, assoclist[i].tx_40mhz ? 40 : 20)
 						: String.format('%.1f <%:Mbit/s%>', assoclist[i].tx_rate / 1000).nobr()
 					;
 				}
@@ -676,14 +691,15 @@
 		<tr class="cbi-section-table-titles">
 			<th class="cbi-section-table-cell">&#160;</th>
 			<th class="cbi-section-table-cell"><%:MAC-Address%></th>
+			<th class="cbi-section-table-cell"><%:Host%></th>
 			<th class="cbi-section-table-cell"><%:Network%></th>
+			<th class="cbi-section-table-cell"><%:Radio%></th>
 			<th class="cbi-section-table-cell"><%:Signal%></th>
-			<th class="cbi-section-table-cell"><%:Noise%></th>
 			<th class="cbi-section-table-cell"><%:RX Rate%></th>
 			<th class="cbi-section-table-cell"><%:TX Rate%></th>
 		</tr>
 		<tr class="cbi-section-table-row">
-			<td colspan="7"><em><br /><%:Collecting data...%></em></td>
+			<td colspan="8"><em><br /><%:Collecting data...%></em></td>
 		</tr>
 	</table>
 </fieldset>
```

**Comment by jow-** [comment]
Instead of truncating the hostname we could put it into a <span> container with `max-width`, `white-space: nowrap`, `overflow: hidden` and `text-overflow: ellipsis`.

Dropping noise column would be fine with me, its not available in many drivers anyway.

**Comment by booo** [comment]
Please keep the noise column.

**Comment by thess** [comment]
I like this but, it would be more useful if this addition could show all associated station's IP and not just the ones using DHCP. A lot my wireless devices have static IP.

**Comment by jow-** [comment]
@booo: Why keep it? It serves no useful purpose and you can still find the value by hovering over the signal indicator icon.

**Comment by jow-** [comment]
@thess: should be doable by mixing in ARP and NDP cache dumps as well.

**Comment by booo** [comment]
@jow- In my day to day activity I compare the signal quite often to the noise. If I have to hover/click each time to see the noise level the interface is far from user friendly.

Maybe someone can post a screenshot or link to the proposed implementation. Maybe the noise level is already a prominent part of the gui and I can't see it in the current screenshot.

**Comment by jow-** [comment]
Since the noise floor is global to the entire radio device and not specific to stations or networks it maybe makes sense to display it somewhere in the header. At least it is redundant to repeat it in each row.

**Comment by booo** [comment]
I agree that the information is redundant in each row.

**Comment by echza** [comment]
how can we enable mauermann's patch without rebuilding the kernel?

**Comment by FreifunkUFO** [comment]
it would be nice to have the "expected throughput" in the table..

**Comment by FreifunkUFO** [comment]
![freifunk-80mhz-](https://cloud.githubusercontent.com/assets/2579265/12196192/0aafa628-b5fd-11e5-8673-d792ed09f299.png)

if there is 80 MHz wifi bandwidth the table is still showing "20 MHz" :-(

**Comment by valentt** [comment]
@FreifunkUFO  I'm seeing 20Mhz when client is resting but when I started download LUCI correctly showed that client switched to 40Mhz channel width.

**Comment by valentt** [comment]
@FreifunkUFO How would you propose to know what is "expected throughput" ?!? Is there some forumula for calculating "expected throughput" ?

**Comment by hnyman** [comment]
@FreifunkUFO @valentt 
The >40 MHz channel width display is problematic. Apparently there is currently no easy way to pass the 80/160 width info.

Based on https://dev.openwrt.org/ticket/20840 I raised that issue as https://github.com/openwrt/luci/issues/533 , and @jow- answered that fixing that requires patching libiwinfo etc., not just Luci.

**Comment by valentt** [comment]
As I openned a https://github.com/openwrt/luci/issues/540 as duplicate ticket, I'll repeat it here, but it is almost identical from OP text:

LUCI  'Associated Stations' list is missing some crucial info, especially for dual band radios.

Currently 'Associated Stations' list is now showing on which radio is client connected, which is really important info to show.

Also showing IP address from ARP or DHCP for each station would be really important info, as should hostname also.

Network column could/should be replaced by Radio name, or should incorporate radio name.

It is common to name both 2.4 and 5.8 network with same SSID if it is on same router, and here LUCI is not usable becaues it doesn't show on which radio which client is connected to.

**Comment by valentt** [comment]
Is it possible to arrange info in two rows? I did a quick mock-up to show this idea.
![stations-two rows](https://cloud.githubusercontent.com/assets/620901/12196963/115e7f52-b603-11e5-9d21-f14f293956e8.png)

**Comment by valentt** [comment]
Also Gargoyle firmware has some station list done quite nicely so we can borrow ideas from them also.
![gargoyle](https://cloud.githubusercontent.com/assets/620901/12197060/02b2e10e-b604-11e5-8677-92cd78089b11.jpeg)

**Comment by valentt** [comment]
Also there are quite a few online admin dashboard templates that we can look for, find good examples and take some elements from them that make sense for LUCI and OpenWrt.
First example: http://rubix305.sketchpixy.com/app/dashboard

**Comment by FreifunkUFO** [comment]
hnyman: ok
valentt: maybe we should use different background colors for these rows, depending on what interface they are? (your problem is only appearing when using same wifi-bssids for different wifi-radios)

about other dashboards: have a look for that brandnew freifunk-gluon status-page https://www.youtube.com/watch?v=lwCwd--Y5R4

Unfortunately that seems to be programmed exclusively for gluon-openwrt (only with centralized "gluon" vpn servers, only with ipv6, you have to use batman-adv and gluon-announce-daemon other needs) 
its only used for getting wifi ad-hoc-statistics but could also be used for Accesspoint-Mode.

**Comment by psyborg55** [comment]
not only OP dropped noise column but SSID column too. both columns should stay while there is still enough space left to add phy/radio column. hostname you can see under DHCP leases and expected throughput is not useful as it's incorrect ->https://github.com/openwrt/mt76/issues/17

**Comment by jow-** [comment]
The expected throughput is incorrect with the mt76 driver but works fine on all others.

**Comment by jow-** [comment]
Disregard the comment above, I thought you meant the phy rate with expected throughput.

**Comment by raz123** [comment]
Wouldn't the best option be to allow one to customize which columns to display?

**Comment by bit2016** [comment]
+1 I think this is an important feature missing. I filter the MACs and its hard to keep track of devices only displaying their MACs

**Comment by jow-** [comment]
I already reworked the assoc list display in 2588364, it looks like this currently:

![assoc](https://cloud.githubusercontent.com/assets/2528802/12639962/bad35194-c5a7-11e5-8021-560ad0d0507a.png)

**Comment by psyborg55** [comment]
so how to build latest trunk without this change?

**Comment by jow-** [comment]
Sigh, whats wrong about this change now?

**Comment by psyborg55** [comment]
it looks ugly with openwrt.org theme

**Comment by juanriccio** [comment]
Recent versions of LuCI do show the hostname and IP address of associated stations, too (nice!), but only when the router is the dhcp provider for that address. I have several routers in AP mode - only the main router has a dhcp server running, so I don't get to see the hostname or IP address.

Could it be possible to query the default (LAN) DNS/DHCP server for the hostname? I guess the IP address could be gathered from local arp tables. If it's too heavy computationally, name lookup could be made optional. I, for one, would welcome that information. It enables me to see at a glance which wireless devices are hooked up to which of the several working routers/APs.

**Comment by LipkeGu** [comment]
@juanriccio you can walkaround this via WINS (samba) slave routers shlould use your main router as WINS Master,

**Comment by bit2016** [comment]
I think it looks great!

But we should be able to manually assign names to the IPs/clients.

That was my first idea.

**Comment by juanriccio** [comment]
@LipkeGu thanks for the suggestion! I set up the main router as a WINS master as detailed for example here: https://wiki.openwrt.org/doc/howto/cifs.server - and I think the setup was successful.

However, I have no idea how to setup the slave routers: they have no Samba service compiled in. That's to keep the image small, I think Samba wouldn't fit. They aren't DHCP clients, either, since they have static IP addresses. How can I set them up to use the master router as a WINS server? Is there any small package that could help? Thank you again.

**Comment by braian87b** [comment]
I think that is better to keep aside the macaddress to "name" translation, data can be retrieved from ethers (dnsmasq dhcp file) from other dns, from wins, or even a custom csv, etc, not neccesary needs to be tied to computer hostname or local dns name... could be a friendly name or device user too.

**Comment by juanriccio** [comment]
Well it would all be fine, as long as I can spot at a glance which wireless devices are connected to that AP. The MAC address alone is not enough for me; I need some symbolic name, or at least an IP address. However, as I said, IP addresses are only shown when the AP is acting as a DHCP server, which in my network just doesn't happen: I have only the main router serving DHCP and DNS. All the APs are just acting as "dumb" APs.

I've had no success with WINS, either, but maybe I'm doing something wrong.

**Comment by FreifunkUFO** [comment]
@juanriccio thats not possible by design (tcp/ip)
the AP has to have some IP interaction with the client to know his IP (but thats not needed for "normal" AP behaviour). it would be possible for an AP without DHCP, f.i. when your AP is in the same IP-Subnet and f.i. ping your client (or whole subnet :-o ) 
in my screenshot the AP is also not the DHCP server (wlan0-2) but somehow AP knows the ipv6 adress of that client..

@jow- would be very nice to have "expected throughput" on the table, even when not working on mt76, would be nice for all others.

@jow- at least the RX/TX rate might be nice when formatting left (not centered!)

![freifunk-assoc-s](https://cloud.githubusercontent.com/assets/2579265/18227345/d8895a68-7221-11e6-97a7-7fe702a10698.png)

**Comment by FreifunkUFO** [comment]
@juanriccio when doing wifi-mesh there was also that problem.. so 9 years ago an extra plugin was written for the routingprotocol olsr, f.i. see https://github.com/servalproject/olsr/blob/master/lib/arprefresh/README_ARPREFRESH

its enabled on my host but in line 2 on the screenshot that doesnt work. but that might be a problem of that plugin or with my meshsetup..

**Comment by juanriccio** [comment]
@FreifunkUFO Thanks for the insight about IP interaction. However, some of my APs do show the IPv6 addresses of associated stations, even if I've gone to great lengths to ensure IPv6 is disabled. Mh, mysteries.

My point is: I do keep names for all station, and the main router has all the naming info I need. The info is available via its DNS/DHCP server, which in turns gets it from /etc/ethers, /etc/hosts and the other usual sources. MACaddr/IPaddr matching can be done via arp and grep on the main router.

I've read the page of the arprefresh plugin that you pointed me to, but as you say it stopped working, and I don't even use olsr, so it's too far out for me. Is there a simple way - a script or something - that can fetch the MACaddr/IPaddr matching info back to the APs?

**Comment by braian87b** [comment]
@juanriccio in order to disable ipv6 assignation on dnsmasq this should be enought:

    uci del dhcp.lan.dhcpv6
    uci del dhcp.lan.ra
    uci commit dhcp; sync
    /etc/init.d/dnsmasq restart

**Comment by braian87b** [comment]
in my system:

`/tmp/dhcp.leases` has all dhcp (static or dynamic) leases, comes from default: `dhcp.@dnsmasq[0].leasefile='/tmp/dhcp.leases'`

`/tmp/etc/dnsmasq.conf` read as `dhcp-host=00:11:22:33:44:57,192.168.1.100,pchostname` for static hosts using `uci add dhcp host` or using luci, lease served or not.

`/etc/ethers` for static hosts (this is the same as `dhcp-host` but just from an external file, lease served or not, just when default `dhcp.@dnsmasq[0].readethers='1'` is `'1'`, could be populated using `echo "00:11:22:33:44:57 192.168.1.100" >> /etc/ethers` this has mac to IP or mac to host, in `/etc/hosts` you could have IP <- hostname, when default `dhcp.@dnsmasq[0].dynamicdhcp='1'` is set to '0' then only host leases in ethers file will be served by dhcp (according to documentation), but actually it serves hosts leases from `/tmp/etc/dnsmasq.conf` with `dhcp-host` too.

I have populated the hosts using uci commands manually, this way it acts I have added using luci.

If i need to I will read the information from `/tmp/dhcp.leases` as described here: http://johnbokma.com/mexit/2008/09/03/dhcp-static-ip-dnsmasq.html

**Comment by juanriccio** [comment]
@braian87b wrote:

> in my system:
> `/tmp/dhcp.leases` [...]
> `/tmp/etc/dnsmasq.conf`

Thanks, braian87b! I know where they are, and I can parse them with
scripts. My issue is that they are on the _main router_ (MR).
Unfortunately, the luci table we are talking about is on the AP.

I know how to deal with this on the processing side (MR), but it's not
enough, because I need to:

1. Have the AP query MR for that info (IPaddr, name) and display it.
2. Make MR listen for such requests or provide data whenever it changes
with reasonable granularity.
3. Extract the info from several system files.

I suppose:

- [ ] 1. Asking for and displaying the info is probably a job for
luci/lua (I've got no clue how to do this).
- [ ] 2. Are there any services I can piggyback on?

- [ ] 3. Actually write the awk/sed/whatever to extract the relevant
info, assuming you are fed with a MACaddr.

I can deal with Task #3.

**Comment by juanriccio** [comment]
As for the side mistery,

@braian87b depicted:

> ![disabling IPv6 in your browser via
luci](https://cloud.githubusercontent.com/assets/6067040/18232056/3d81a17e-729e-11e6-864c-de2b5c901e59.png)

I'd done this already, that's why I found it annoying! :rage:

**Comment by braian87b** [comment]
@juanriccio Sorry, I didn't read correctly what you described, I interpreted something else... I actually managed to get info on AP from MR.

**Comment by braian87b** [comment]
opkg update
	opkg install luci-mod-rpc # brings luci-lib-json
	opkg install rpcd-mod-file # brings zlib libopenssl libcurl
	opkg install curl

	/etc/init.d/uhttpd restart
	/etc/init.d/rcpd restart
	# if not works, then reboot:
	ubus list | grep file # check, now should list `file`


	# assing permissions (beware this will add full permissions to rcpd access, you should set as you need
	cat<<'EOF' > /usr/share/rpcd/acl.d/superuser.json
	{
	   "superuser": {
	      "description": "Super user access role",
	      "read": {
	         "ubus": {
	           "*": [ "*" ]
	         },
	         "uci": [ "*" ]
	      },
	      "write": {
	         "ubus": {
	          "*": [ "*" ]
	         },
	         "uci": [ "*" ]
	      }
	   }
	}
	EOF



	rm /tmp/test_output.txt
	IP_HOST_MR=192.168.1.1
	PASSWORD=toor
	. /usr/share/libubox/jshn.sh
	JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "00000000000000000000000000000000", "session", "login", { "username": "root", "password": "'$PASSWORD'"  } ] }' http://$IP_HOST_MR/ubus)
	json_load "$JSON"
	json_select result
	json_select 2
	json_get_var SESION_ID ubus_rpc_session
	JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "'$SESION_ID'", "file", "read", { "path": "/tmp/dhcp.leases" } ] }' http://$IP_HOST_MR/ubus)
	json_load "$JSON"
	json_select result
	json_select 2
	json_get_var FILE_DATA data
	echo "$FILE_DATA" > /tmp/test_output.txt
	cat /tmp/test_output.txt


Sorry again for misunderstanding...

**Comment by braian87b** [comment]
There is a simple way, just using arp and a simple reverse DNS query, but I don't know if will work for you... `arp -a` should always retrieve _all_ the active hosts on the current AP

	root@OpenWrt:~# arp -a
	IP address       HW type     Flags       HW address            Mask     Device
	192.168.1.34     0x1         0x2         ec:08:6b:xx:xx:xx     *        br-lan
	192.168.1.35     0x1         0x2         84:d6:d0:xx:xx:xx     *        br-lan
		
	root@APAluYBraian:~# arp -a | grep 84:d6:d0:xx:xx:xx
	192.168.9.35     0x1         0x2         84:d6:d0:xx:xx:xx     *        br-lan

	root@OpenWrt:~# nslookup 192.168.1.35  192.168.1.1
	Server:    192.168.1.1
	Address 1: 192.168.1.1 mainrouter
	Name:      192.168.1.35
	Address 1: 192.168.1.35 **KindleFire.lan**

	# keeping this in mind, we can do:
	MAC_TO_QUERY=84:d6:d0:xx:xx:xx
	IP_QUERY=$(arp -a | grep $MAC_TO_QUERY | cut -d' ' -f1)
	HOSTNAME_QUERY=$(nslookup `arp -a | grep $MAC_TO_QUERY | cut -d' ' -f1` 127.0.0.1 | tail -n 1 | grep -o '[^ ]*$' | cut -d' ' -f 1)
	echo $IP_QUERY
	echo $HOSTNAME_QUERY

**Comment by braian87b** [comment]
I have the CC 15.05.1 version of OpenWrt, I don't know how to install a newer version of luci without break anything (please if any of you know, tell me), but I have tested the _DHCP Leases_ table, if I replace the `/tmp/dhcp.leases` and obviusly it takes the data from there, maybe the _Associated Stations_ table do the same.

**Comment by juanriccio** [comment]
@braian87b  Great replies, thanks!

> There is a simple way, just using arp and a simple reverse DNS query, but I don't know if will work for you...  arp -a  should always retrieve all the active hosts on the current AP

On my AP systems, the `arp` command doesn't report the IP address of associated stations, since as @FreifunkUFO wrote earlier, there's no IP interaction at all, and everything happens at Layer 2. Only the MR knows about IP addresses and symbolic names.

I've cooked up a couple of scripts to be run on the MR via the default `ash` shell. Both take a single parameter - the MAC address of an associated station - and return respectively the IP address and the hostname.

mac2ip:

	#!/bin/ash
	# mac2ip - return the IP address of an associated station given its MAC address
	
	mac2ipaddr()
	{
		grep -i $1 /var/dhcp.leases \
		| head -1 \
		| cut -d' ' -f3
	}
	
	if [ "$#" -eq 1 ]; then
		mac2ipaddr $1
	fi


mac2host:

	#!/bin/ash
	# mac2host - return the hostname of an associated station given its MAC address
	
	mac2hostname()
	{
		grep -i $1 /var/dhcp.leases \
		| head -1 \
		| cut -d' ' -f4
	}
	
	if [ "$#" -eq 1 ]; then
		mac2hostname $1
	fi


However, my knowledge of rpc/JSON mechanisms is null. Assuming you have such scripts available and functional on the MR, how would you go about calling them from the AP?

> I have the CC 15.05.1 version of OpenWrt, I don't know how to install a newer version of luci without break anything (please if any of you know, tell me), but I have tested the DHCP Leases table, if I replace the  /tmp/dhcp.leases  and obviusly it takes the data from there, maybe the Associated Stations table do the same.

I don't understand - do you mean trouble with patching the rpc/JSON statements into the AP's luci page again? I guess it shouldn't take much to any knowledgeable guy to fix the luci/lua code. Maybe someone will pick up from there...?  :wink:

**Comment by juanriccio** [comment]
@braian87b wrote:

> It is not too pretty but it actually works:
> 
> Run this on some AP and point  IP_HOST_MR  and  PASSWORD  vars to your main Router
> 
>   opkg update
>   opkg install luci-mod-rpc # brings luci-lib-json
>   opkg install rpcd-mod-file # brings zlib libopenssl libcurl
>   opkg install curl

How about the main router? I installed the same packages on the MR, but this code fails on the AP.

> ```
>   # assing permissions (beware this will add full permissions to rcpd access, you should set as you need
>   cat<<'EOF' > /usr/share/rpcd/acl.d/superuser.json
>   {
>      "superuser": {
>         "description": "Super user access role",
>         "read": {
>            "ubus": {
>              "*": [ "*" ]
>            },
>            "uci": [ "*" ]
>         },
>         "write": {
>            "ubus": {
>             "*": [ "*" ]
>            },
>            "uci": [ "*" ]
>         }
>      }
>   }
>   EOF
> 
>   rm /tmp/test_output.txt
>   IP_HOST_MR=192.168.1.1
>   PASSWORD=toor
>   . /usr/share/libubox/jshn.sh
>   JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "00000000000000000000000000000000", "session", "login", { "username": "root", "password": "'$PASSWORD'"  } ] }' http://$IP_HOST_MR/ubus)
>   json_load "$JSON"
>   json_select result
>   json_select 2
>   json_get_var SESION_ID ubus_rpc_session
>   JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "'$SESION_ID'", "file", "read", { "path": "/tmp/dhcp.leases" } ] }' http://$IP_HOST_MR/ubus)
>   json_load "$JSON"
>   json_select result
>   json_select 2
>   json_get_var FILE_DATA data
>   echo "$FILE_DATA" > /tmp/test_output.txt
>   rm /usr/share/rpcd/acl.d/superuser.json
> ```

Here's what I get:

    # ./getleases.sh
    WARNING: Variable 'result' does not exist or is not an array/object
    WARNING: Variable '2' does not exist or is not an array/object
    
    # wc /tmp/test_output.txt
            1         0         1 /tmp/test_output.txt
    #

**Comment by braian87b** [comment]
@juanriccio 
Sorry, yesterday was about 1:30am I didn't express myself well, please read again the edit post: https://github.com/openwrt/luci/issues/540#issuecomment-244657293
I think I was more clear now.

I mean that I have the ChaosCalmer 15.05.1 version of OpenWrt from https://downloads.openwrt.org/chaos_calmer/15.05.1/ar71xx/generic/ , I don't know how to install a newer version of luci to test the new **Associated Stations** table and see if it works or not.
I don't know noting about luci actually, I use more terminal than luci... but I want to be able to test it too but I don't know how.

Are this new features present on https://downloads.openwrt.org/snapshots/trunk/ar71xx/generic/ ?

**Comment by juanriccio** [comment]
@braian87b 
> read again the edited post

now I see: it's the other way round! Now I see. I could have figured it out myself. Thanks for straightening it up; I'm going to play with it tomorrow.

> I use more terminal than luci... but I want to be able to test it too but I don't know how.
> Are this new features present on https://downloads.openwrt.org/snapshots/trunk/ar71xx/generic/ ?

I think so, at least if you pick a version with luci. Otherwise, you can install luci via `opkg` - as you did for, say, `curl` - and start it by issuing something like
```
opkg update && opkg install luci
/etc/init.d/uhttpd start
```

**Comment by braian87b** [comment]
@juanriccio  Yess, well, I will sysupgrade to snapshots/trunk to see if this is there.

**Comment by juanriccio** [comment]
I found some time to try and play with this ideas, and I came up with a script to be run on the main router. Here it is.

```
#!/bin/sh

# includes the functions that we need to parse JSON files
. /usr/share/libubox/jshn.sh

case "$1" in
    list)
        echo '{ "mac2host": { "macaddr": "macaddr" }, "mac2ip": { "macaddr": "macaddr"} }'
    ;;
    call)
        case "$2" in
            mac2host)
                # read the argument
                read input
                # Load the argument into the json for retrieval
                json_load "$input"
                json_get_var mac "macaddr"
                hostname=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f4)
                echo "{ \"hostname\":\"$hostname\" }"
            ;;
            mac2ip)
                # read the argument
                read input
                # Load the argument into the json for retrieval
                json_load "$input"
                json_get_var mac "macaddr"
                ipaddr=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f3)
                echo "{ \"ipaddr\" :\"$ipaddr\" }"
            ;;
        esac
    ;;
esac
```
I placed it in the MR's /usr/libexec/rpcd directory, and it does work locally. 
```
# ubus call mactranslate mac2host  '{"macaddr": "00:11:22:33:44:55"}'
{"hostname":"android-123456abcd123456"}
```
How do I setup things on the AP's side? I need to incorporate the rpc call and resulting data into luci. I also need to set minimal permissions.

**Comment by braian87b** [comment]
Run this only once on Main Router (to enable rpc calling from dumb AP's):

	opkg update
	opkg install luci-mod-rpc # brings luci-lib-json

	/etc/init.d/uhttpd restart
	/etc/init.d/rcpd restart
	# if not works, then reboot:
	ubus list | grep mac # check, should list `mactranslate` or `mac2host`

This too, its needed assing permissions (beware this will add full permissions to rcpd access, you should set as you need (read documentation to learn how)

	cat<<'EOF' > /usr/share/rpcd/acl.d/superuser.json
	{
	   "superuser": {
	      "description": "Super user access role",
	      "read": {
	         "ubus": {
	           "*": [ "*" ]
	         },
	         "uci": [ "*" ]
	      },
	      "write": {
	         "ubus": {
	          "*": [ "*" ]
	         },
	         "uci": [ "*" ]
	      }
	   }
	}
	EOF

Run this on some AP and point `IP_HOST_MR` and `PASSWORD` vars to your main Router

Run this only once:

	opkg update
	opkg install curl

Copy and paste this to test

	rm /tmp/test_output.txt
	IP_HOST_MR=192.168.1.1
	PASSWORD=toor
	. /usr/share/libubox/jshn.sh
	JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "00000000000000000000000000000000", "session", "login", { "username": "root", "password": "'$PASSWORD'"  } ] }' http://$IP_HOST_MR/ubus)
	json_load "$JSON"
	json_select result
	json_select 2
	json_get_var SESION_ID ubus_rpc_session
	JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "'$SESION_ID'", "mactranslate", "mac2host", { "macaddr": "00:11:22:33:44:55" } ] }' http://$IP_HOST_MR/ubus)
	json_load "$JSON"
	json_select result
	json_select 2
	json_get_var FILE_DATA data
	echo "$FILE_DATA" > /tmp/test_output.txt
	cat /tmp/test_output.txt

`/tmp/test_output.txt` contents (should say this if it worked fine):

    {"hostname":"android-123456abcd123456"}

Please let me know about how it worked.

**Comment by juanriccio** [comment]
Thanks, @braian87b!
The scripts worked already. My question was mainly about luci integration and about the minimal permission thing. However, I guess I can run this with full permissions on the server (MR) side, so now it's probably just a matter of integrating and packaging.

Here are the scripts. On the MR:

```
#!/bin/sh

# mactranslate - script for master router
# examines dhcp leases
# and returns hostname and IP addr for a given macaddr

# includes the functions that we need to parse JSON files
. /usr/share/libubox/jshn.sh

case "$1" in
    list)
        echo '{ "mac2host": { "macaddr": "macaddr" }, "mac2ip": { "macaddr": "macaddr"}, "mac2full": {"macaddr": "macaddr"} }'
    ;;
    call)
        case "$2" in
            mac2host)
                # read the argument
                read input
                # Load the argument into the json for retrieval
                json_load "$input"
                json_get_var mac "macaddr"
                hostname=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f4)
                echo "{ \"data\":\"$hostname\" }"
            ;;
            mac2ip)
                # read the argument
                read input
                # Load the argument into the json for retrieval
                json_load "$input"
                json_get_var mac "macaddr"
                ipaddr=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f3)
                echo "{ \"data\":\"$ipaddr\" }"
            ;;
            mac2full)
                # read the argument
                read input
                # Load the argument into the json for retrieval
                json_load "$input"
                json_get_var mac "macaddr"
                hostname=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f4)
                ipaddr=$(grep -i "$mac" /var/dhcp.leases \
                    | head -1 \
                    | cut -d' ' -f3)
                echo "{ \"data\":\"$hostname ($ipaddr)\" }"
            ;;
        esac
    ;;
esac
```

On the AP:
```
# getmacinfo.sh - AP side script
# gets hostname and IP addr for a given macaddr from master router

usage=$(cat <<'END_HEREDOC'
getmacinfo.sh [options] <mac_addr>
options:
 -h  return hostname
 -i  return ipaddress
 -a  return "hostname (ipaddress)"
default: -a
END_HEREDOC
)

IP_HOST_MR=192.168.1.1          # the master router IP address
PASSWORD=toor                   # root password on master rouer
METHOD="mac2full"               # default: "hostname (ipaddress)" string

while getopts ":ahi" opt; do
  case $opt in
    a)
      METHOD="mac2full"
      ;;
    h)
      METHOD="mac2host"
      ;;
    i)
      METHOD="mac2ip"
      ;;
    \?)
      echo "$0 invalid option: -$OPTARG" >&2
      ;;
  esac
done
shift $((OPTIND-1))
if [ $# -ne 1 ]; then
        echo "$usage"
        exit 1
fi

. /usr/share/libubox/jshn.sh

JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "00000000000000000000000000000000", "session", "login", { "username": "root", "password": "'$PASSWORD'"  } ] }' http://$IP_HOST_MR/ubus)
json_load "$JSON"
json_select result
json_select 2
json_get_var SESSION_ID ubus_rpc_session
JSON=$(curl -s -d '{ "jsonrpc": "2.0", "id": 1, "method": "call", "params": [ "'$SESSION_ID'", "mactranslate", "'$METHOD'", { "macaddr": "'$1'" } ] }' http://$IP_HOST_MR/ubus)
json_load "$JSON"
json_select result
json_select 2
json_get_var FILE_DATA data
echo "$FILE_DATA"

```

A sample incorrect call (performed on the AP):
```
# ./getmacinfo.sh
getmacinfo.sh [options] <mac_addr>
options:
 -h  return hostname
 -i  return ipaddress
 -a  return "hostname (ipaddress)"
default: -a

```

A correct call, again on the AP:
```
# ./getmacinfo.sh 00:11:22:33:44:55
android-123456abcd123456 (192.168.1.52)

```

All that's needed now is to get this bit of information in the "Associated Stations" table :smile:

**Comment by braian87b** [comment]
This is amazing, I think now we should gain the atention of a experience developer of Luci to integrate this code and make it work... having this should be really easy to get done, may be this should be added to a package and we could get an experience package dev to help us.

It is really common to need to know the actual device when seeing the Web UI, and more than common to have a Main Router and several dumb AP's... should be great too to get on the Main Router info about in which one of many dumb AP's is connected which device... what do you think? ;)

**Comment by braian87b** [comment]
I found this, here is more info about ACL's: https://wiki.openwrt.org/doc/techref/ubus#acls

**Comment by juanriccio** [comment]
@braian87b wrote:
> should be great too to get on the Main Router info about in which one of many dumb AP's is connected which device... what do you think? ;)

I agree - however, it's not as simple. I can't think of a way to do this efficiently, without lots of querying or automatic, timed notifications. The matter is I don't know the system, in particular all the system hooks that probably are already in place and could make this an easy task.

> I found this, here is more info about ACL's: https://wiki.openwrt.org/doc/techref/ubus#acls

Aha, thanks! I'll read that doc. Hopefully, I'll figure out how the ACL business works and possibly update the scripts accordingly.

**Comment by cshoredaniel** [comment]
I think this (MAC -> address translation) would be a useful feature, but getting it work will take some significant time and testing.  One issue is that you have to handle the case where main router is not a version with OpenWrt/LEDE that can communicate the addresses to the the AP's, and on networks without IP addresses (i.e. pure bridges) the AP won't have the information most of the time.

**Comment by braian87b** [comment]
Hi, in my opinion its not a problem on "networks without IP addresses / pure bridges"...
actually the several test on scripts that @juanriccio and me performed is more than sufficient to gather the information on dumb AP's and send it to _(or get if from)_ the Main Router _(the one with dhcp)_ and save it somewhere _(we could have a plaintext tab-delimited or sqlite, is more than enough)_ and then in luci check if that plaintext or sqlite is available and then show as an extra column or instead of macaddress (the macaddres could remain in a tooltip) or even put it as a tooltip or even in a separate new page.

I think that what just we need is to create a init.d script (to act as a pseudo daemon and update the info when a new host appears, and put a minimal uci config and then put it all together in a new package and make it available, I can promise to try to get some init.d work and publish here the final files, but as we previusly talked about with @juanriccio some messages ago # **we need someone** who knows both, how to put the info in luci, convert into a package and with access to publish as package in the official repositories.
imagine that would be something like:

**in the Main Router:** 

    opkg update, opkg install luci-macaddr-to-name macaddr-to-name

 _(one for adding luci info page, and one for daemon that updates the tab-delimited or sqlite db)_

**on the Dumb AP's** _(or in the main router too if it acts as AP too)_: 

    opkg update, opkg install macaddr-to-name-ap

What do you think?

**Comment by cshoredaniel** [comment]
@braian87b Unfortunately, I think I may have some rain on this optimism:

1) Assumes both 'dumb' AP and main router are both running on versions or LEDE/OpenWrt with package included.
2) Security & stability: how exactly are you transferring these files, ensuring integrity etc.?
3) Really the transfer part probably belongs in packages feed
4) Need someone with time/resources to put in this who has (or can acquire) the necessary skills.

Thoughts?

**Comment by braian87b** [comment]
Since today it's Friday, and I do not usually be ironic, but yesterday I saw some videos of Linus Torvals and I just started thinking about how, maybe, he would answer this message, I hope do not take this as offense, it's just to add a little incentive, I am already too happy with OpenWrt and do not need more, but as I am happy I just want to give my time to help to improve it my own 'grain of salt'. 👍  :) 

1. and left out people who has all routers openwrt capable in their facilities with full access to them.
2. we should encript all in case some spy actually starts decrypting our wpa2 or even get physical acces to an ethernet jack, he could get sensitive info that he could gain just using nmap, arp and sniffing.
3. that what I said, if info is available show it on luci, if not then crash the entire luci interface or the radio too, even kernel panic the router, because put `if` instructions is too much for this feature
4. this is opensource... just people that has no free spare time and no desire of get this working for himselfs will be the people tortured until this new feature is finished.

**Comment by cshoredaniel** [comment]
@braian87b I'm not entirely sure what you mean by some of what you've said, but:

1) True, but that means it's not a useful as a more general solution
2) Erm, actually there are wireless attacks that work, depending on certain factors, and most users don't know what factors matter.
3) Optional is fine, however see 4)
4) I don't know about others working on LuCI but I am not paid to develop this, nor do I have any obligation except my personal motivations to work on this, and I personally already have plans for working on a number of issues and PRs from the community that I consider higher priority.  However, this is opensource, if you can't do it yourself, hire someone with the skills to do it, and submit a PR with an Apache License (v2) and it'll get reviewed and have either requests for appropriate changes, or integrated (assuming there are enough active developers to be able to do that; if not, new blood is needed or the project will die).

**Comment by cshoredaniel** [comment]
2) Oh as for nmap and such, point taken, however that is not my real concern; my concern is how dumb router accesses the data (I haven't read your code, besides I'd like to hear you explain the design), or the main router pushes the data to the dumb router.  Also where is the main router storing the file?  It'd be bad to write to limited write flash for this, and if it's on e.g. /tmp, how does it get from there to the dump ap?

**Comment by cshoredaniel** [comment]
@braian87b Sorry missed your ubus calls and script earlier - it's a long thread and I missed parts (only recently started looking at outstanding issues).

**Comment by cshoredaniel** [comment]
@braian87b I have concerns about the implied use of passwords with access to the main router (what permissions and how is that managed) on every dumb router.

**Comment by braian87b** [comment]
Sorry for the misunderstanding...
1. people who wants this feature are free to contribute if they want to have it in their routers and make the same benefit to others that apply the same (if people have other routers then should not be worried about)
2. we should start with something simple and working... many implementations have no security involved and then are secured with a 3rd party or even in a ssl tunnel compiled version etc. actually we can just use nc (ncat) as a plain simple text pseudoprotocol many things started like that.
3. as optional can be outside of luci main package and be in other one, with their specific luci page.
4. what I mean is that I am not demanding time from anyone, just asked if someone wants to contribute, just someone that has knowledge, spare time, and desire to have this feature for himselft to (if someone doesn't meet this all three requierements then probably will feel this as a no-confortable thing to do)...
Sorry @cshoredaniel I missed that you are contributor of luci, I didn't pay attention to that until now.

As I actually misunderstood the main -> ap, ap -> main info sent/get... I think that if someone from luci can create a text file with the info and luci package that shows that then other (me and others) can make what is needed to populate that file... 
for starting, can populate that file using /tmp/dhcp.leases and get from there, macaddress, ip, hostname from that file, then we could get static ip/host from dhcp.@host[x], /etc/ethers and later with computers that have actually configured in-place a static ip then will have to use a arp or reverse dns query or `nmblookup -A 192.168.1.100` (that returns the hostname using netbios query).. or fill a list or file manually, etc, etc. This is because I insist that we should get this info from a new dedicated file, because will be more easily to fill that file from several places instead of add to too much overload in luci reading data from many sources or reading incomplete data from just one source...
As this info changes from time to time I think that should not be saved on firmware, should be on /tmp and if someone wants to persist it change the path to extroot or another mounted filesystem.

If all of you are keeping interesting in this I could take some time to think more about a proper and better design to get this idea more complete.

**Comment by cshoredaniel** [comment]
@braian87b Sorry for misunderstanding your request as a push for work.  I still have concerns about the security aspect and am not a fan of the 'start insecure, we'll figure out the security later'; I like to at least know it's not a permanent hole.  The problem with /tmp is that is is not accessible from /www which is where uhttpd accessible files live and I'm not keen on the ubus mechanism you describe.  If someone disabled symlinks you can't symlink from /www to /tmp either.  I think there are some challenges you haven't yet considered.  Even though it's technically gatherable data I'm not convinced that means unauthenticated access is a good idea, or access via plaintext (or obfuscated but plaintext when used) password on devices is wise.  At the very least this should easily turned off, and be done using nobody:nogroup for the service providing the informaiton (i.e. unprivileged user).

**Comment by braian87b** [comment]
@cshoredaniel Interesting, I didn't know the /tmp <-> /www limitation I will keep in mind that. I don't want to push people to work, just ask for anyone who wants to contribute, but just if they want to, and they are actually interested in this. The data that will travel now will have a list of ip<>macaddress(<>hostname) that info could be read only by devices actually inside the network, and if some device is already inside the network can actually read and get same data using tools like arp and nmap... as I said, if someone here has knowledge in C programming I think it could easily write a daemon with ssl and problem solved... I think reading the data could not be a problem, the problem is when in a more serious implementation can happen something like bad people faking the data (the data will just be presented to a user in luci) but don't know how important will actually be in that implementation, I think this is worse actually. Is like to have luci using http/80 and have to configure it to use https/443... We first need to get a first version of the idea working and if it works well and the testers of the trunk branch are interested in having the feature probably more people will want to contribute patching and improving the code and even rewrite the entire code to C with security in mind, etc... The ubus mechanism was just a "test", actually that day I was testing that because I wanted to know if it was possible to get info from an OpenWRT box without using ssh, was just a coincidence same day that I found this conversation. We should leave this idea here, hope someone would be capable in the future to improve the idea and describe it better than me. I am not expert on OpenWRT, but since I use it I want to contribute.
Thanks for the interest! :) 👍

**Comment by braian87b** [comment]
Hehe, if you read my first message on this topic, that message have my only main concern about this topic (actually the later data interchange thing was just a night experiment playing with a openwrt box at hand) 
https://github.com/openwrt/luci/issues/540#issuecomment-244476594
I think that is the thing that we need to talk about and decide if is better to have it in separate file, or read from multiple files... What do you thing ?

**Comment by braian87b** [comment]
I think we should delete our coments and put the relevant info in just simplified in just one message, all these comments just added noise to this conversation.

**Comment by cshoredaniel** [comment]
@braian87b Could you open a separate issue with only the relevant bits for that one topic?  That doesn't seem like something you'd be blocked on.

**Comment by cshoredaniel** [comment]
@braian87b actually can you move those bits to #724

**Comment by psyborg55** [comment]
this is more useful information, when you can have overview of all interfaces in one screen. rest is just bullshit


<a href='https://postimg.org/image/wcbjccqex/' target='_blank'><img src='https://s28.postimg.org/t5gzsq5z1/Screenshot_1.png' border='0' alt='Screenshot-1'/><br /><a target='_blank' href='https://postimage.org/app.php'>hydrogenerator dnp2</a><br /><br />

**Comment by braian87b** [comment]
@psyborg55 please expland your explanation, which information are you saying that is more useful?

**Comment by FrancYescO** [comment]
i recently installed OpenWRT on my WDR3600 but i cannot understand why i cannot sees this upgrade, hostnames/ips are still missing in the Associated Stations list

Firmware Version | OpenWrt Chaos Calmer 15.05.1 / LuCI 15.05-149-g0d8bbd2 Release (git-15.363.78009-956be55)

opkg update / upkg list-upgradable show nothing... is there something that i have to do to upgrade LuCI?

**Comment by braian87b** [comment]
@jow-  Why did you close this ?

**Comment by jow-** [comment]
The original request has been solved (additional info is now there). The discussion later seems to be a feature request nobody is working on. I do not plan to implement inter-access-point device queries.

**Comment by oavaldezi** [comment]
It seems there is a desire to have the hostnames displayed in the "Associated Stations" section. Could this be read from /etc/dnsmasq.conf or /etc/config/dhcp ?

Both those files can contain records for MAC and IP addresses and hostnames.

**Comment by jow-** [comment]
This already happens. The issue is with dump AP setups which do not act as DHCP server or router themselves. Such devices will never have L3 connectivity with their wireless clients and thus no info about associated IP addresses or host names.

**Comment by braian87b** [comment]
The main problem relies that for humans (the ones that uses Luci Web UI) there is no correlation to mac address, so, I will be useful if we can have some sort of known devices list that Luci can use it to translate the mac address to a user defined name... The hostname translation it could be an initial data fill. But it will be better if user can define each one by itself...

so, if you could have
"xx:xx:xx..." -> "Bob iPhone"
"xx:xx:xy..." -> "Alice Notebook"
"xx:xx:xz..." -> "Alice Samsung Galaxy S9"
"xx:xx:yy..." -> "Alice iPad"
"xx:xx:yz..." -> "Bob Thinkpad"
in the way:
```
xx:xx:xx...\tBob iPhone\n
xx:xx:xy..\tAlice Notebook\n
xx:xx:xz...\tAlice Samsung Galaxy S9\n
xx:xx:yy...\tAlice iPad\n
xx:xx:yz...\tBob Thinkpad\n
```

We just NEED that some Luci Developer could make the necessary to read some file at /tmp/macaddresses_aliases and do that translation if the macaddress is found on that list (and leave the mac address between parenthesis or in hover tooltip, and do for DHCP leases and for stations at least. If the sysadmin fills that file with whatever script they want...

It is this difficult to do? please It would be REALLY USEFUL!

**Comment by jow-** [comment]
You can achieve that already by declaring static DHCP leases without IP addresses.

**Comment by braian87b** [comment]
No, sorry but we don't want to get involve in this dnsmasq or other OpenWRT component or even deal with defining hostname that could convert later in problems.

### We just need a plain simple macaddress-to-human translation

**Comment by jow-** [comment]
Then just use `/etc/ethers`. Simply supply a name instead of an IP address for the MAC.

**Comment by braian87b** [comment]
Buahh, no, no IP, mac addresses... .. forget it... leave it as it is now, no problem...

**Comment by jow-** [comment]
You wanted a file in the form `xx:xx:xx\tBob iPhone\n` this is *exactly* what `/etc/ethers` is. First column mac, second column hostname *or* IP.

**Comment by braian87b** [comment]
/etc/ethers is a network system file, we want a file that just involves Luci, and nothing else. so we can play with it using scripting, etc.

**Comment by jow-** [comment]
Then use the workaround to declare a static DHCP lease *without* IP address in `/etc/config/dhcp`:

```
config host
	option name 'fairphone'
	option mac '6c:ad:f8:43:40:fb'
```

Such entries are ignored by dnsmasq and recognized by LuCI as name hints for MAC addresses.

**Comment by braian87b** [comment]
what is a "hint" it wil replace the macaddress by the defined "name" on Luci ?

**Comment by jow-** [comment]
![hints](https://user-images.githubusercontent.com/2528802/43478728-f59eafde-94fe-11e8-997b-cf248d44dd0d.png)

**Comment by cshoredaniel** [comment]
Luci isn't getting huge amounts of dev at this time--not too many people
*volunteering* to work on it (i.e. people with the skill offering their
time and effort for free), and no one seem to wants things badly enough
to hire someone to get it done and submit to OpenWrt's LuCI feed.

For *volunteers* who actually care about the GUI, I think there's little
desire to work on old (current) LuCI because the goal is move to
luci-ng; but AFAIK luci-ng isn't far enough along for folks to use it as
the main GUI, and then build their favourite app on top.

My 2 cents.

**Comment by thess** [comment]
@jow - cool hack. I didn't know I could do that. Solves an old problem of mine identifying some old 11b devices on my home network. THANKS

**Comment by cshoredaniel** [comment]
Oh and I do plan on helping the situation by getting to the effort that was on hold while I was ill, of splitting up LuCI and having some sort of LuCI + luci-ng hybrid to smooth the transition.  I'm hoping that a) work and b) make it easier to transition instead of having this chicken and egg problem.

**Comment by braian87b** [comment]
@cshoredaniel Thanks for that info, I didn't knew that.

**Comment by oavaldezi** [comment]
I have records such as this one in /etc/dnsmasq.conf, but they don't work as hints for Luci:

dhcp-host=b8:27:eb:34:cc:2d,192.168.0.40,raspberrypi1

Is it because Luci doesn't read /etc/dnsmasq.conf (only /etc/config/dhcp), or is it because they contain IP addresses?

**Comment by braian87b** [comment]
I insist for one last time, Luci should read a different file for that only purpose and each user should fill it in the way they want, even If i put on ethers file dnsmasq will read it and I dont want that.
I will find now luci-ng too see how it is going, hope for best, seems no luck here.

**Comment by oavaldezi** [comment]
That's what you want, and the way you want it. I don't mind using /etc/dnsmasq.conf, and as a last resort, I won't mind using /etc/config/dhcp.

**Comment by cshoredaniel** [comment]
@oavaldezi Luci doesn't read custom dnsmasq.conf. (Lua UCI is where Luci names comes from; dnsmsaq.conf is not uci format and _most_ of Luci revolves around uci configurations (there are exceptions, especially in newer code, and especially for status information).  Anyway dnsmasq.conf is not how Luci is built to understand/use dnsmsaq (except by way of generating .conf files from uci configuration, but it's the uci configuration that is understood, not the .conf). HTH.

**Comment by braian87b** [comment]
I think that I'll give a try: https://github.com/zhaojh329/vuci

**Comment by cshoredaniel** [comment]
@jow- @hnyman FYI I've started on the 'low-hanging fruit' for moving things into separate apps instead of all in mod-admin-full and such, and as it happens I picked on the the hostnames definition.  Ironically, except for the fact that they live in /etc/config/dhcp there is nothing about the definitions that is specific to dnsmasq (or unbound, or odhcpd, etc), which I found interesting.   @EricLuehrsen might be especially interested in this since I thing for unbound the only thing he picks up from /etc/config/dhcp is the hostnames (as DHCP becomes a separately handled entity). 

@EricLuehrsen is there anything your unbound work uses besides hostnames from /etc/config/dhcp?

**Comment by oavaldezi** [comment]
Thank you @cshoredaniel . Your answer made me read about UCI, and I learned that its configuration files are kept in `/etc/config/`. That precludes using `/etc/dnsmasq.conf`.

I'll try the workaround suggested above by @jow- : to declare a static DHCP lease without IP address in `/etc/config/dhcp`

**Comment by cshoredaniel** [comment]
@jow I am wondering if there is a way to preserve the translation work when making these changes (it'd not be so good to lose translations because of moving things around).

**Comment by EricLuehrsen** [comment]
@cshoredaniel Unbound UCI attempts to mimic dnsmasq UCI features OpenWrt community is accustom to. It uses `/etc/config/dhcp` to figure out which interfaces are DHCP (LAN) and which are not (WAN). It optionally will snoop for configured pairing with dnsmasq. It will optionally pull the `domain`, `mx`, `srv`, and `cname` records originally intended for dnsmasq. But as for base DHCP, why wouldn't predetermined host names be in that file? That would seem to be a natural location.

**Comment by cshoredaniel** [comment]
@EricLuehrsen to clarify: at present my changes shouldn't be 'user-visible'.  They are aimed at being 'under-the-hood' AND don't change the underlying UCI configs.  The idea is to decouple the various parts of the UI so that one an pick and choose.

**Comment by cshoredaniel** [comment]
@EricLuehrsen @jow- @hnyman You'll see what I mean with the move of the 'Hostnames' page to luci-app-hostnames from luci-mod-admin-full.   The idea is that luci-mod-admin-full is (IMO) a tar baby at the moment, and needs to get less sticky.

**Comment by jow-** [comment]
@cshoredaniel - before you start digging deeper into it, can you please outline what you're exactly trying to do (in a separate issue or RFC PR)?

I'm not overly happy into breaking luci-mod-admin-full into tiny pieces. I've been fighting for single kilobytes in last months and breaking it down into lots of micro packages will add back a lot of packaging bloat. For example with something like `luci-app-hostnames` the packaging overhead can easily by three to four times the actual code size.

**Comment by cshoredaniel** [comment]
@jow- ah, hadn't realized about the packagin; that is a definite issue...definitely needs another way to achieve the same goal.  At one time I had created https://github.com/openwrt/luci/issues/869; I will reopen that an retitle.  BTW is there any way to pull this discussion (the relevant bits) in to there?

**Comment by braian87b** [comment]
Hi @jow- 

I tested successfully your solution and it could work for me for the Dumb AP's, but I still had a problem with the dhcp leases table on the main router.
```
# first way:
# Delete all dhcp.@host's ('While until error')
while uci del dhcp.@host[-1] 2> /dev/null; do :; done
uci add dhcp host
uci set dhcp.@host[-1].name='phone_bob'
uci set dhcp.@host[-1].mac='aa:77:14:xx:xx:xx'
uci add dhcp host
uci set dhcp.@host[-1].name='table_alice'
uci set dhcp.@host[-1].mac='ec:aa:fd:xx:xx:xx'
uci add dhcp host
uci set dhcp.@host[-1].name='notebook_bob'
uci set dhcp.@host[-1].mac='44:2c:aa:xx:xx:xx'
uci commit dhcp

# the other way: (it has to be space not tab, and does not allow spaces on name):
cat<<'EOF' > /etc/ethers
aa:77:14:xx:xx:xx phone_bob
ec:aa:fd:xx:xx:xx table_alice
44:2c:aa:xx:xx:xx notebook_bob
EOF

# If both are present will shown "first_way (second_way)""

# Delete everything:
while uci del dhcp.@host[-1] 2> /dev/null; do :; done
uci commit dhcp
cat "" > /etc/ethers
```

No problems so far for Associated Stations on dumb AP's
Is there some way to see the device names on the dhcp leases on the MainAP without having nothing to do with the hostname (since hostname is a LAN internal DNS)... ??? I really don't want to modify or allow to be host resolved or even discovered on the LAN those names. It is just for internal usage only on LuCI and nowhere else.

That's is reason I wanted to have an individual file for this human macaddress-to-device_name translation.

We could even use that list on many other places, like in "Status->Routes->ARP" where it shows the ARP list of known devices and shows on which interface it could be found... (that is really useful for a NetAdmin/SysAdmin).

On Status->SystemLog, we could pipe the log to a sed/awk replace script and have the Actual device names instead of the mac addresses there too it will be useful, really, a lot!.

We have something similar In Status->Realtime Graphs->Connections there we had a nice IP-to-hostname translations if we wait a few seconds.

Thanks.
